### ISSUE TITLE
`==` on collections containing `missing` breaks its contract

### ISSUE THREAD

**Issue by Seelengrab** [JuliaLang/julia#52484]
> help?> ==
> search: == === !== = .= >= => <= !=
>
>   ==(x, y)
>[...]
> For collections, missing is returned if at least one of the
>  operands contains a missing value and all non-missing values are equal. Use isequal or === to always get a Bool result.

Yet:

```
julia> (1,missing) == (missing,2)
missing

julia> [1,missing] == [missing,2]
missing
```

Per the docstring, these should both be `false`, since the non-missing values are not equal.

**Comment by nalimilan** [comment]
Maybe the docs could be made more explicit and say "all non-missing pairs of values".

**Comment by Seelengrab** [comment]
To me, the docstring implies usage of a `skipmissing`-like behavior, i.e. "in the absence of `missing`, are the two equal?". If that is `true`, and there is at least one `missing` in either collection, then the result is `missing`. If there is no `missing`, the result is whatever the former returns.

**Comment by adienes** [comment]
I would expect it to be the same as `all(x .== y)` so I think the current behavior is correct

**Comment by Seelengrab** [comment]
> And at least I can assure you that it behaves as intended. ;-) 

Then the docstring isn't precise enough in the behavior that is guaranteed, since the docstring is at the very least ambiguous in what is required/guaranteed.

> Existing code can rely on the implementation, but not on an interpretation of the docstring which doesn't match the actual behavior.

That seems backwards to me - relying on what is _documented_ is what has always been the go-to in Julia ("if it's not in the Manual, it's not API"), so hearing this is directly counter to what has been practiced so far. To me it feels like you're trying to justify the existing behavior post-fact.

**Comment by adienes** [comment]
my reading of the existing documentation matches the existing behavior

> since the docstring is at the very least ambiguous

I would call it at the very most ambiguous? maybe the intersection of those two bounds makes it exactly equal ambiguous.

I certainly wouldn't agree that it's worded in a way _counter_ to existing behavior

**Comment by Seelengrab** [comment]
If swapping the order in the conditional changes the result (that is, checking for "all non-missing values are equal" first), the documentation is certainly ambiguous. You're reading the docstring like `return_missing = has_missing || all_different`, whereas I'm reading it like `return_missing = has_missing && all_different`, because of that `and` there, suggesting a conjunction instead of a disjunction. I don't know how you get to that disjunction when the docstring clearly says `and`.

**Comment by mbauman** [comment]
The "all non-missing values are equal" clause could actually be read in at least _three_ ways:
* Every single non-missing value on both LHS and RHS must be `==` to eachother.  In other words, both sides must only have two unique values: `missing` and some `x`.
* After applying `skipmissing` to each argument **independently**, the resulting missing-less collections are iterated jointly and the iterated pairs of values must be equal.
* After applying `Missings.skipmissings` to the arguments **jointly**, the resulting missing-less collections are iterated jointly and the iterated pairs of values must be equal.

Docs do double-duty: they must be both accessible to newcomers while still giving technical guidance to advanced users.

**Comment by mbauman** [comment]
Right, which is precisely why this isn't a "contract" that should force implementation changes counter to the very much intended design.

**Comment by Seelengrab** [comment]
So what can I as a user that wants to implement `==` on a type refer to as a "ground truth"? What properties do I _need_ to obey? "Check the implementation" just isn't feasible when there isn't necessarily an applicable reference implementation to compare to.

**Comment by nalimilan** [comment]
Maybe the simplest way to document this would be by mentioning the result of calling `==` on each pair of values? Something like:
> For collections, `missing` is returned if calling `==` on each pair of values in matching positions
in `x` and `y` returns at least one `missing` and no `false`. For arrays, this is equivalent to `all(x .== y)`.

**Comment by Seelengrab** [comment]
What does "pair of values in matching positions" mean for non-orderable collections like `Set`?

**Comment by rfourquet** [comment]
> I'd argue that this NOT being missing is correct

I think that's intended, because `Set` membership is defined by `isequal`, and `isequal(missing, missing)` is true. Of course `Set` is a "collection", but doesn't fit in the `==` docstring from the OP, whose wording should be amended accordingly.

**Comment by Seelengrab** [comment]
Yes, that's (partially) my point - `==` propagating missing or not depends on the object, not just "it's a collection" or "it wraps other objects". For the same reason, `Some(missing) == Some(1)` should be `false`, not `missing`.
